remotes/fixup.go

```diff
@@ -136,3 +177,128 @@ func checkBaseImage(baseImage *bundle.BaseImage) error {
 
 	return nil
 }
+
+// FixupEvent is an event that is raised by the Fixup Logic
+type FixupEvent struct {
+	SourceImage    string
+	DestinationRef reference.Named
+	EventType      FixupEventType
+	Message        string
+	Error          error
+	Progress       ProgressSnapshot
+}
+
+// FixupEventType is the the type of event raised by the Fixup logic
+type FixupEventType string
+
+const (
+	// FixupEventTypeCopyImageStart is raised when the Fixup logic starts copying an
+	// image
+	FixupEventTypeCopyImageStart = FixupEventType("CopyImageStart")
+
+	// FixupEventTypeCopyImageEnd is raised when the Fixup logic stops copying an
+	// image. Error might be populated
+	FixupEventTypeCopyImageEnd = FixupEventType("CopyImageEnd")
+
+	// FixupEventTypeProgress is raised when Fixup logic reports progression
+	FixupEventTypeProgress = FixupEventType("Progress")
+)
+
+type descriptorProgress struct {
+	ocischemav1.Descriptor
+	done     bool
+	action   string
+	err      error
+	children []*descriptorProgress
+	mut      sync.Mutex
+}
+
+func (p *descriptorProgress) markDone() {
+	p.mut.Lock()
+	defer p.mut.Unlock()
+	p.done = true
+}
+
+func (p *descriptorProgress) setAction(a string) {
+	p.mut.Lock()
+	defer p.mut.Unlock()
+	p.action = a
+}
+
+func (p *descriptorProgress) setError(err error) {
+	p.mut.Lock()
+	defer p.mut.Unlock()
+	p.err = err
+}
+
+func (p *descriptorProgress) addChild(child *descriptorProgress) {
+	p.mut.Lock()
+	defer p.mut.Unlock()
+	p.children = append(p.children, child)
+}
+
+func (p *descriptorProgress) snapshot() DescriptorProgressSnapshot {
+	p.mut.Lock()
+	defer p.mut.Unlock()
+	result := DescriptorProgressSnapshot{
+		Descriptor: p.Descriptor,
+		Done:       p.done,
+		Action:     p.action,
+		Error:      p.err,
+	}
+	if len(p.children) != 0 {
+		result.Children = make([]DescriptorProgressSnapshot, len(p.children))
+		for ix, child := range p.children {
+			result.Children[ix] = child.snapshot()
+		}
+	}
+	return result
+}
+
+type progress struct {
+	roots    []*descriptorProgress
+	walkDone bool
+	mut      sync.Mutex
+}
+
+func (p *progress) addRoot(root *descriptorProgress) {
+	p.mut.Lock()
+	defer p.mut.Unlock()
+	p.roots = append(p.roots, root)
+}
+
+func (p *progress) markWalkDone() {
+	p.mut.Lock()
+	defer p.mut.Unlock()
+	p.walkDone = true
+}
+
+func (p *progress) snapshot() ProgressSnapshot {
+	p.mut.Lock()
+	defer p.mut.Unlock()
+	result := ProgressSnapshot{
+		WalkDone: p.walkDone,
+	}
+	if len(p.roots) != 0 {
+		result.Roots = make([]DescriptorProgressSnapshot, len(p.roots))
+		for ix, root := range p.roots {
+			result.Roots[ix] = root.snapshot()
+		}
+	}
+	return result
+}
+
+// DescriptorProgressSnapshot describes the current progress of a descriptor
```

Comment:
Eventually under the hood there is a `containerd/content.Status` associated with the `Pusher`, which contains offet and total level of details so you could do %ages for the push (I think?).

Might be worth putting that on some list (mental or otherwise) as a useful future enhancement, since pushing a large layer might take a while.

Reply:
Yes, we could do that (having a goroutine that snapshots the progress every second or so reporting upload progress when we have cross-registry fixups)